# Jaya Vijayen
## Winter 2022 Data Science Intern Challenge

### Question 1: Given some sample data, write a program to answer the following: click here to access the required data set.

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

1. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 

2. What metric would you report for this dataset?

3. What is its value?


## Load and Inspect the Dataset

In [1]:
# load packages for this analysis
import csv
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import datetime
import matplotlib as mpl
import matplotlib.pyplot as plt

In [2]:
# dataset was given by Shopify
# load data into a Pandas dataframe and read the csv
# name the Pandas dataframe as shopify_df


shopify_df = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv', keep_default_na=True)

In [3]:
# Return the shape to see how much data we have to work with from this CSV file

shopify_df.shape

(5000, 7)

There are 5000 rows and 7 columns in this CSV file

In [4]:
# Show the count and data types for all columns

shopify_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
order_id          5000 non-null int64
shop_id           5000 non-null int64
user_id           5000 non-null int64
order_amount      5000 non-null int64
total_items       5000 non-null int64
payment_method    5000 non-null object
created_at        5000 non-null object
dtypes: int64(5), object(2)
memory usage: 273.5+ KB


The dataset above seem to be complete.  At the moment, all the data columns has values(5000 entries in all the columns).  There are no missing values.

In [5]:
# inspect the top rows of this data using the .head function

shopify_df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

The .head() function shows the data without having to load the entire dataset. It reveals the first 5 rows of the dataset. It shows all the 'columns' for each order. The .head() function is extremely useful when I want to scan the first few rows of a very large dataset.

In [6]:
# inspect the bottom rows of this data using the .tail function

shopify_df.tail()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18

The .tail() function shows the data in the last 5 rows of the dataset. It shows all the 'columns' for each order starting from the bottom of the dataset. Again, the .tail() function is extremely useful when I want to scan the last few rows of a very large dataset.

In [7]:
# Let's check if there are any duplicate values.

sum(shopify_df.duplicated())

0

There are no duplicate records.

## Exploratory Data Analysis (EDA)

In [8]:
# Here are some statistics using pd.describe

shopify_df.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [9]:
# Descriptive statistics on the order amount column

shopify_df['order_amount'].describe()

count      5000.000000
mean       3145.128000
std       41282.539349
min          90.000000
25%         163.000000
50%         284.000000
75%         390.000000
max      704000.000000
Name: order_amount, dtype: float64

In [10]:
# the average order value(AOV) is also the mean
# total revenue divided by total number of orders

shopify_df['order_amount'].mean()

3145.128

In [11]:
# the median

shopify_df['order_amount'].median()

284.0

The above describe() method shows the AOV is indeed 3145.13 as mentioned in Question 1. we now know how this number is derived. We also can see that there is a large standard deviation of 41282.54. On average the values vary 41282.54 from the mean. In addition, the minimum value is 90 and the maximum value is 704000.  Also, we can the values for the median(284), first quartile and the second quartile.  With the all this information, we can see the maximum value is much higher than median, first quartile and the second quartile. This seem to indicate that the AOV is not a very useful metric.  

In [14]:
# the mode

mode = stats.mode(shopify_df['order_amount'])
print("The mode value is {} with a count of {}".format(mode.mode[0], mode.count[0]))

The mode value is 153 with a count of 87


__a. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.__

From the results above the average order value(AOV) or mean of 3145.13 is way higher then the median(284) and the mode(153).  The reason why the mean is so incredibly high is because the largest order value is 704000.  This large number(704000) alone has skewed the AOV way to high. This indicates that the AOV gives us a partial picture of the customers buying behavior. In this case, AOV was misleading. So it is always a good practice to evaluate the data with other measures of central tendency like median and mode.


__b. What metric would you report for this dataset?__ 

From the statistics above, we know that 50% of all orders are lower than 284. The mode which is the most common dollar amount of  orders is 153. So the mode or the most frequent order value is a good metric to report for this dataset.

__c. What is its value?__ 

$153.00

### Question 2: For this question you’ll need to use SQL. Follow this link(https://www.w3schools.com/SQL/TRYSQL.ASP?FILENAME=TRYSQL_SELECT_ALL) to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below. 

__a.	How many orders were shipped by Speedy Express in total?__

SELECT COUNT (OrderID) FROM Orders  
JOIN Shippers ON Orders.ShipperID = Shippers.ShipperID  
WHERE Shippers.ShipperName = 'Speedy Express'  

Answer is **_54_**

__b.	What is the last name of the employee with the most orders?__

SELECT COUNT(OrderID), Employees.LastName FROM Orders  
JOIN Employees ON Orders.EmployeeID = Employees.EmployeeID  
GROUP BY Employees.LastName  
ORDER BY COUNT(OrderID) DESC  

Answer is **_Peacock with 40 orders_**

__c.	What product was ordered the most by customers in Germany?__

SELECT Products.ProductName, SUM(OrderDetails.Quantity) AS "TotalOrdered"  
FROM Orders  
JOIN Customers ON Customers.CustomerID = Orders.CustomerID  
JOIN OrderDetails ON OrderDetails.OrderID = Orders.OrderID  
JOIN Products ON Products.ProductID = OrderDetails.ProductID  
WHERE Customers.Country = 'Germany'  
GROUP BY OrderDetails.ProductID  
ORDER BY TotalOrdered DESC  

Answer is **_Boston Crab Meat with 160 total orders_**